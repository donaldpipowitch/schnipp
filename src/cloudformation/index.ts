import {
  CloudFormationCustomResourceEvent,
  CloudFormationCustomResourceResponse
} from 'aws-lambda';
import { IHandler } from '../';

export type CloudFormationCustomResourceHandler = IHandler<
  CloudFormationCustomResourceEvent,
  CloudFormationCustomResourceResponse
>;
