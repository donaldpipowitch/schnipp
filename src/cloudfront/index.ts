import {
  CloudFrontRequestEvent,
  CloudFrontRequestResult,
  CloudFrontResponseEvent,
  CloudFrontResponseResult
} from 'aws-lambda';
import { IHandler } from '..';

export type CloudFrontRequestHandler = IHandler<
  CloudFrontRequestEvent,
  CloudFrontRequestResult
>;

export type CloudFrontResponseHandler = IHandler<
  CloudFrontResponseEvent,
  CloudFrontResponseResult
>;
