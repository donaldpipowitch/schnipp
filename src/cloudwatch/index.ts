import { CloudWatchLogsEvent } from 'aws-lambda';
import { IHandler } from '../';

export type CloudwatchHandler = IHandler<CloudWatchLogsEvent, void>;
