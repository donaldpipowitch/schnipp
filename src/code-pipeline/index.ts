import { IHandler } from '..';
import { CodePipelineEvent } from 'aws-lambda';

export type CodePipelineHandler = IHandler<CodePipelineEvent>;
