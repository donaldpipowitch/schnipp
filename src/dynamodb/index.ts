import { DynamoDBStreamEvent } from 'aws-lambda';
import { IHandler } from '../';

export type DynamoDBStreamHandler = IHandler<DynamoDBStreamEvent, void>;
