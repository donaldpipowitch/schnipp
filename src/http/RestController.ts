import { APIGatewayHandler, createResponse } from '.';
import { CONTAINER_INSTANCE_PROP, Container, Newable } from 'simple-ts-di';
import {
  APIGatewayEvent,
  Context,
  APIGatewayProxyResult,
  APIGatewayProxyEvent
} from 'aws-lambda';
import {
  findMethod,
  MethodMetaData,
  DecoratedExceptionFilter,
  ExceptionFilter,
  Guard,
  MetaData
} from './factories/createHttpMethodDecorator';
import { HttpException } from './exceptions/HttpException';
import { ForbiddenException } from './exceptions/ForbiddenException';
import { getControllerMetaData } from './factories/createClassDectorator';

export class RestController implements APIGatewayHandler {
  private controllerData = getControllerMetaData(this.constructor);

  private get container() {
    return (this as any)[CONTAINER_INSTANCE_PROP] as Container;
  }

  async handle(
    event: APIGatewayProxyEvent,
    context: Context
  ): Promise<APIGatewayProxyResult> {
    const method = findMethod(
      this.controllerData.basePath,
      event.resource,
      event.httpMethod,
      this
    );

    if (method == null) {
      return createResponse(404);
    }

    const guards = [...method.guards, ...this.controllerData.guards];
    const metaData = { ...this.controllerData.metaData, ...method.metaData };

    try {
      await this.processGuards(guards, event, context, metaData);

      const headers = { ...this.controllerData.headers, ...method.headers };

      return createResponse(
        method.statusCode!,
        await this.generateResponse(method, event, context),
        headers
      );
    } catch (exception) {
      const filters = [...method.filters, ...this.controllerData.filters];

      const mappedExceptionResponse = await this.processFilters(
        filters,
        exception,
        event,
        context,
        metaData
      );

      if (mappedExceptionResponse != null) {
        return mappedExceptionResponse;
      }

      if (exception instanceof HttpException) {
        return createResponse(exception.statusCode, {
          statusCode: exception.statusCode,
          message: exception.message
        });
      }

      return createResponse(500, {
        statusCode: 500,
        message: 'Internal Server Error'
      });
    }
  }

  private async generateResponse(
    method: MethodMetaData,
    event: APIGatewayEvent,
    context: Context
  ) {
    return await (this as any)[method.propertyKey].apply(
      this,
      method.paramaters
        .sort((a, b) => a.index - b.index)
        .map(r => r.injector(event, method.metaData, context))
    );
  }

  private async processFilters(
    filters: Array<DecoratedExceptionFilter>,
    e: any,
    event: APIGatewayEvent,
    context: Context,
    metaData: MetaData
  ) {
    try {
      const exceptionFilter = filters.find(x =>
        x.exceptions.some(y => e.name === y.name || e instanceof y)
      );

      if (exceptionFilter == null) {
        return;
      }

      const filter = await this.container.get<ExceptionFilter>(
        exceptionFilter.filter
      );

      return await filter.catch(e, event, metaData, context);
    } catch (e) {
      return createResponse(500, { message: 'Internal Server Error' });
    }
  }

  private async processGuards(
    guards: Array<Newable<Guard>>,
    event: APIGatewayEvent,
    context: Context,
    metaData: MetaData
  ) {
    const canActivate = (await Promise.all(
      guards.map(guard =>
        this.container
          .get(guard)
          .then(guard => guard.canActivate(event, metaData, context))
      )
    )).every(Boolean);

    if (!canActivate) {
      throw new ForbiddenException();
    }
  }
}
