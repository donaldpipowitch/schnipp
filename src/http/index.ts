import {
  APIGatewayEvent,
  APIGatewayProxyResult,
  CustomAuthorizerEvent,
  CustomAuthorizerResult
} from 'aws-lambda';
import { IHandler } from '..';

export type APIGatewayHandler = IHandler<
  APIGatewayEvent,
  APIGatewayProxyResult
>;

export type AuthorizeHandler = IHandler<
  CustomAuthorizerEvent,
  CustomAuthorizerResult
>;

export { RestController } from './RestController';

export {
  InjectedParam,
  MetaData,
  MethodMetaData,
  ParamInjector,
  DecoratedExceptionFilter,
  Guard,
  ExceptionFilter
} from './factories/createHttpMethodDecorator';

export { Catch } from './decorators/catch';
export {
  GET,
  POST,
  DELETE,
  PUT,
  PATCH,
  OPTIONS,
  HEAD
} from './decorators/methods';
export { Path } from './decorators/path';
export { Param } from './decorators/param';
export { Query } from './decorators/query';
export { Header } from './decorators/header';
export { StatusCode } from './decorators/statusCode';
export { AuthContext } from './decorators/authContext';
export { Meta } from './decorators/meta';
export { Body } from './decorators/body';
export { Use } from './decorators/use';

export { HttpException } from './exceptions/HttpException';
export { NotFoundException } from './exceptions/NotFoundException';
export { ForbiddenException } from './exceptions/ForbiddenException';

export { AuthPolicy, HttpVerb, Effect } from './utils/AuthPolicy';
export { createResponse } from './utils/createResponse';
