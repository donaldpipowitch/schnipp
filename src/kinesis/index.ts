import { KinesisStreamEvent } from 'aws-lambda';
import { IHandler } from '..';

export type KinesisStreamHandler = IHandler<KinesisStreamEvent>;
