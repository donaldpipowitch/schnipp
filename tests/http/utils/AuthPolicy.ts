import { AuthPolicy, HttpVerb } from '../../../src/http';

describe('AuthPolicy', () => {
  it('should parse ARN', () => {
    expect(
      AuthPolicy.parseMethodArn(
        'arn:aws:execute-api:us-east-1:random-account-id:random-api-id/dev/GET/events/*'
      )
    ).toMatchSnapshot();
  });

  it('should let me create an valid poliy document (allow)', () => {
    const o = AuthPolicy.parseMethodArn(
      'arn:aws:execute-api:us-east-1:random-account-id:random-api-id/dev/GET/events/*'
    );
    const p = new AuthPolicy('me', o.accountId, o);

    p.allowMethod(HttpVerb.DELETE, '/bar');
    p.allowMethodWithConditions(HttpVerb.DELETE, '/bar2', {
      cond: { foo: 'bar' }
    });

    p.addContext({
      foo: 'bar'
    });

    expect(p.build()).toMatchSnapshot();
  });

  it('should let me create an valid poliy document (deny)', () => {
    const o = AuthPolicy.parseMethodArn(
      'arn:aws:execute-api:us-east-1:random-account-id:random-api-id/dev/GET/events/*'
    );
    const p = new AuthPolicy('me', o.accountId, o);

    p.denyMethod(HttpVerb.DELETE, '/bar');
    p.denyMethodWithConditions(HttpVerb.DELETE, '/bar2', {
      cond: { foo: 'bar' }
    });

    p.addContext({
      foo: 'bar'
    });

    expect(p.build()).toMatchSnapshot();
  });

  it('should let me create an valid poliy document denyAll', () => {
    const o = AuthPolicy.parseMethodArn(
      'arn:aws:execute-api:us-east-1:random-account-id:random-api-id/dev/GET/events/*'
    );
    const p = new AuthPolicy('me', o.accountId, o);

    p.denyAllMethods();

    p.addContext({
      foo: 'bar'
    });

    expect(p.build()).toMatchSnapshot();
  });

  it('should let me create an valid poliy document allowAll', () => {
    const o = AuthPolicy.parseMethodArn(
      'arn:aws:execute-api:us-east-1:random-account-id:random-api-id/dev/GET/events/*'
    );
    const p = new AuthPolicy('me', o.accountId, o);

    p.allowAllMethods();

    p.addContext({
      foo: 'bar'
    });

    expect(p.build()).toMatchSnapshot();
  });

  it('should fail on empty policy', () => {
    const p = new AuthPolicy('foo', 'bar');
    expect(() => p.build()).toThrow();
  });

  it('should fail on invalid path', () => {
    const p = new AuthPolicy('foo', 'bar');
    expect(() => p.allowMethod(HttpVerb.GET, '$%&/&%$ยง"')).toThrow();
  });

  it('should work for only allow or deny', () => {
    const p = new AuthPolicy('foo', 'bar');
    p.allowMethod(HttpVerb.GET, '/');
    expect(p.build()).toMatchSnapshot();

    const p2 = new AuthPolicy('foo', 'bar');
    p2.denyMethod(HttpVerb.GET, '/');
    expect(p2.build()).toMatchSnapshot();
  });
});
