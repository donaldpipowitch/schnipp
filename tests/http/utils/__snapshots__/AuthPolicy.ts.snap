// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthPolicy should let me create an valid poliy document (allow) 1`] = `
Object {
  "context": Object {
    "foo": "bar",
  },
  "policyDocument": Object {
    "Statement": Array [
      Object {
        "Action": "execute-api:Invoke",
        "Condition": Object {
          "cond": Object {
            "foo": "bar",
          },
        },
        "Effect": "Allow",
        "Resource": Array [
          "arn:aws:execute-api:us-east-1:random-account-id:random-api-id/dev/DELETE/bar2",
        ],
      },
      Object {
        "Action": "execute-api:Invoke",
        "Effect": "Allow",
        "Resource": Array [
          "arn:aws:execute-api:us-east-1:random-account-id:random-api-id/dev/DELETE/bar",
        ],
      },
    ],
    "Version": "2012-10-17",
  },
  "principalId": "me",
}
`;

exports[`AuthPolicy should let me create an valid poliy document (deny) 1`] = `
Object {
  "context": Object {
    "foo": "bar",
  },
  "policyDocument": Object {
    "Statement": Array [
      Object {
        "Action": "execute-api:Invoke",
        "Condition": Object {
          "cond": Object {
            "foo": "bar",
          },
        },
        "Effect": "Deny",
        "Resource": Array [
          "arn:aws:execute-api:us-east-1:random-account-id:random-api-id/dev/DELETE/bar2",
        ],
      },
      Object {
        "Action": "execute-api:Invoke",
        "Effect": "Deny",
        "Resource": Array [
          "arn:aws:execute-api:us-east-1:random-account-id:random-api-id/dev/DELETE/bar",
        ],
      },
    ],
    "Version": "2012-10-17",
  },
  "principalId": "me",
}
`;

exports[`AuthPolicy should let me create an valid poliy document allowAll 1`] = `
Object {
  "context": Object {
    "foo": "bar",
  },
  "policyDocument": Object {
    "Statement": Array [
      Object {
        "Action": "execute-api:Invoke",
        "Effect": "Allow",
        "Resource": Array [
          "arn:aws:execute-api:us-east-1:random-account-id:random-api-id/dev/*/*",
        ],
      },
    ],
    "Version": "2012-10-17",
  },
  "principalId": "me",
}
`;

exports[`AuthPolicy should let me create an valid poliy document denyAll 1`] = `
Object {
  "context": Object {
    "foo": "bar",
  },
  "policyDocument": Object {
    "Statement": Array [
      Object {
        "Action": "execute-api:Invoke",
        "Effect": "Deny",
        "Resource": Array [
          "arn:aws:execute-api:us-east-1:random-account-id:random-api-id/dev/*/*",
        ],
      },
    ],
    "Version": "2012-10-17",
  },
  "principalId": "me",
}
`;

exports[`AuthPolicy should parse ARN 1`] = `
Object {
  "accountId": "random-account-id",
  "method": "GET",
  "region": "us-east-1",
  "restApiId": "random-api-id",
  "stage": "dev",
}
`;

exports[`AuthPolicy should work for only allow or deny 1`] = `
Object {
  "context": Object {},
  "policyDocument": Object {
    "Statement": Array [
      Object {
        "Action": "execute-api:Invoke",
        "Effect": "Allow",
        "Resource": Array [
          "arn:aws:execute-api:*:bar:*/*/GET/",
        ],
      },
    ],
    "Version": "2012-10-17",
  },
  "principalId": "foo",
}
`;

exports[`AuthPolicy should work for only allow or deny 2`] = `
Object {
  "context": Object {},
  "policyDocument": Object {
    "Statement": Array [
      Object {
        "Action": "execute-api:Invoke",
        "Effect": "Deny",
        "Resource": Array [
          "arn:aws:execute-api:*:bar:*/*/GET/",
        ],
      },
    ],
    "Version": "2012-10-17",
  },
  "principalId": "foo",
}
`;
